src/common/function/src/scalars/geo/geohash.rs

```diff
@@ -17,22 +17,21 @@ use std::sync::Arc;
 
 use common_error::ext::{BoxedError, PlainError};
 use common_error::status_code::StatusCode;
-use common_query::error::{self, InvalidFuncArgsSnafu, Result};
-use datafusion::arrow::datatypes::Field;
+use common_query::error::{self, Result};
+use datafusion::arrow::array::{Array, AsArray, ListBuilder, StringViewBuilder};
+use datafusion::arrow::datatypes::{DataType, Field, Float64Type, UInt8Type};
+use datafusion::logical_expr::ColumnarValue;
+use datafusion_common::utils;
 use datafusion_expr::type_coercion::aggregates::INTEGERS;
-use datafusion_expr::{Signature, TypeSignature, Volatility};
-use datatypes::arrow::datatypes::DataType;
-use datatypes::prelude::ConcreteDataType;
-use datatypes::scalars::{Scalar, ScalarVectorBuilder};
-use datatypes::value::{ListValue, Value};
-use datatypes::vectors::{ListVectorBuilder, MutableVector, StringVectorBuilder, VectorRef};
+use datafusion_expr::{ScalarFunctionArgs, Signature, TypeSignature, Volatility};
 use geohash::Coord;
-use snafu::{ResultExt, ensure};
+use snafu::ResultExt;
 
-use crate::function::{Function, FunctionContext};
+use crate::function::Function;
+use crate::scalars::geo::helpers;
 
 macro_rules! ensure_resolution_usize {
-    ($v: ident) => {
+    ($v: expr) => {
```

Comment:
using `expr` here can be dangerous, the expression might be evaluated multiple times for each `$v`. I suggest to keep it `ident`, if we have an expr for this, we can do `let v = expr;` first.